text/0628-prettier.md

```diff
@@ -0,0 +1,135 @@
+- Start Date: 2020-05-18
+- Relevant Team(s): Ember CLI
+- RFC PR: https://github.com/emberjs/rfcs/pull/628
+- Tracking: (leave this empty)
+
+# Add Prettier
+
+## Summary
+
+This RFC proposes adding [Prettier](https://prettier.io/) to the blueprints
+that back `ember new` and `ember addon`.
+
+## Motivation
+
+Using Prettier removes the ever present and ongoing stylistic debates that tend
+to happen on teams. Prettier is incredibly freeing for both developers **and**
+reviewers of a codebase. The developer can author in whatever format they find
+easiest to type out, avoiding needless wasted time trying to get that
+indentation _just_ right, and either setup an automated "format on save"
+operation in their editor or use a quick command line invocation to `--fix` all
+files at once.
+
+This dove tails very nicely with Embers deeply ingrained goal of using shared
+solutions to solve common problems, and aligns very well with the general trend
+to play better with the general JavaScript ecosystem.
+
+## Detailed design
+
+Due to the pre-existing design (and improvements added to the blueprints in
+[ember-cli/rfcs#121](https://github.com/emberjs/rfcs/blob/master/text/0121-remove-ember-cli-eslint.md))
+implementation is very straight forward. The general idea is that we will
+update the `app` and `addon` blueprints to add a few Prettier related packages
+to the `package.json`, update the linting configuration to enforce Prettier
+styles, and update the other blueprint code to ensure that its output is
+Prettier compatible.
+
+### Packages
+
+The following packages will be added to the `package.json` of both `app` and `addon` blueprints:
+
+* [prettier](https://www.npmjs.com/package/prettier) - The main package (used by the other packages).
+* [eslint-config-prettier](https://www.npmjs.com/package/eslint-config-prettier) - Disables stylistic ESLint rules that would conflict with the Prettier output.
+* [eslint-plugin-prettier](https://www.npmjs.com/package/eslint-plugin-prettier) - Enables `eslint-config-prettier`s `recommended` set, and adds the `prettier/prettier` rule to enforce styles.
+
+### Configuration Changes
+
+#### ESLint
+
+The `.eslintrc.js` that is generated will be updated to add:
+
+```js
+{
+  extends: ['plugin:prettier/recommended'],
+}
+```
+
+#### Prettier
+
+A new `prettier` section will be added to the `package.json` file with the following contents:
+
+```json
+{
+  "prettier": {
+    "singleQuotes": true
+  }
+}
+```
+
+There is a [small number of configuration
+options](https://prettier.io/docs/en/options.html) that can be used in this
+file, but we are intentionally avoiding modifying the default values for things
+that are not _nearly_ universally agreed on in the Ember ecosystem.
```

Comment:
I believe some of agreements exist in [editorconfig](https://github.com/ember-cli/ember-new-output/blob/master/.editorconfig) today. Are we going to respect editorconfig settings with prettier, or should we consider removing editorconfig at some point?

Reply:
> Are we going to respect editorconfig settings with prettier?

Yep, Prettier already parses and uses the `.editorconfig` values for some of its defaults. It is a bit buried, but see [these docs](https://prettier.io/docs/en/api.html#prettierresolveconfigfilepath--options):

> If options.editorconfig is true and an .editorconfig file is in your project, Prettier will parse it and convert its properties to the corresponding prettier configuration. This configuration will be overridden by .prettierrc, etc. Currently, the following EditorConfig properties are supported:
> 
> * end_of_line
> * indent_style
> * indent_size/tab_width
> * max_line_length